### PR TITLE
[MAINTENANCE] Support Adapter Model ".safetensors" File Format During HF Upload

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* This change updates Ludwig to handle changes introduced in HuggingFace PEFT 0.7.0: Adapter weights are now stored in the ".safetensors" format, instead of the PyTorch default Pickle ".bin" format.  This change caused Ludwig to error when uploading the trained model to HuggingFace, because Ludwig verifies acceptable file extensions.  With this change, both are allowed for now; however, there is a "TODO" to formalize the deprecation policy, and subsequently discontinue supporting ".bin" format.
* The new test module for testing upload functionality is created and contains the initial tests for the new model file format.
* Enhancements in type hinting and improvements to exception types in "upload_utils.py" is also included with this change.

### Remarks
* Please note that even though HuggingFace PEFT now uses "safetensors" as the default, it is backward-compatible, meaning that "adapter_mode.bin" files, uploaded via the earlier versions of PEFT, will still be read-in correctly (please see "https://github.com/huggingface/peft/blob/e73967edea0b337fb11fdde3da5b9823512f8f68/src/peft/utils/save_and_load.py#L236" for details):
```
    if use_safetensors:
        adapters_weights = safe_load_file(filename, device=device)
    else:
        adapters_weights = torch.load(filename, map_location=torch.device(device))
```

# Code Pull Requests

Please provide the following:

- a clear explanation of what your code does
- if applicable, a reference to an issue
- a reproducible test for your PR (code, config and data sample)

# Documentation Pull Requests

Note that the documentation HTML files are in `docs/` while the Markdown sources are in `mkdocs/docs`.

If you are proposing a modification to the documentation you should change only the Markdown files.

`api.md` is automatically generated from the docstrings in the code, so if you want to change something in that file, first modify `ludwig/api.py` docstring, then run `mkdocs/code_docs_autogen.py`, which will create `mkdocs/docs/api.md` .

### CODE REVIEW

**Review by arnavgarg1 — APPROVED** [review]
PR looks great! Thanks for adding this backwards compatibility check and also adding a nice test set for this :)

File: tests/ludwig/utils/test_upload_utils.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+@pytest.mark.unit
+def test_upload_to_hf_hub__validate_upload_parameters(
+    output_directory_manager, file_names: list[str], error_raised: tuple[type, str] | None
+):
+    """Test "HuggingFaceHub._validate_upload_parameters()", which is executed in the path of upload to HuggingFace
+    Hub; for example: `upload hf_hub -repo_id "hf-account/repo-name" --model_path.
+
+    /content/results/api_experiment_run`.
+
+    Eeach test case consists of: 1) Populating the temporary output directory ("training_results_output) with zero or
```
Comment (justinxzhao) [inline tests/ludwig/utils/test_upload_utils.py:137]: nit: Each

**Review by justinxzhao — APPROVED** [review]

File: tests/ludwig/utils/test_upload_utils.py (hunk 1 above)
Comment (alexsherstinsky) [reply tests/ludwig/utils/test_upload_utils.py]: Nice catch -- fixed -- thank you!
